File: Frontend/src/services/dhikrService.js
/**
 * Dhikr/Dua API Service
 * Handles all dhikr and dua-related API calls
 */

import apiClient from '../utils/apiClient';

/**
 * Get a random dhikr/dua for daily spiritual practice
 * @param {number} [categoryId=1] - Category of dhikr (morning, evening, general, etc.)
 * @returns {Promise<Object>} Random dhikr from the specified category
 */
export async function getDailyDhikr(categoryId = 1) {
  try {
    if (categoryId < 1) {
      throw new Error('Invalid category ID. Must be greater than 0.');
    }

    const response = await apiClient.get('/api/dhikr/daily', {
      params: { category_id: categoryId }
    });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get all dhikr/dua from a specific category
 * @param {number} categoryId - Category ID to filter by
 * @param {number} [limit=20] - Maximum results to return (max 100)
 * @param {number} [offset=0] - Pagination offset
 * @returns {Promise<Object>} List of dhikr from the category
 */
export async function getDhikrByCategory(categoryId, limit = 20, offset = 0) {
  try {
    if (!categoryId || categoryId < 1) {
      throw new Error('Invalid category ID. Must be greater than 0.');
    }
    if (limit > 100) {
      throw new Error('Limit cannot exceed 100.');
    }
    if (offset < 0) {
      throw new Error('Offset must be non-negative.');
    }

    const response = await apiClient.get(`/api/dhikr/category/${categoryId}`, {
      params: { limit, offset }
    });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get a random dhikr/dua, optionally filtered by category
 * @param {number} [categoryId] - Optional category to filter random selection
 * @returns {Promise<Object>} Random dhikr
 */
export async function getRandomDhikr(categoryId) {
  try {
    const params = {};
    if (categoryId) {
      if (categoryId < 1) {
        throw new Error('Invalid category ID. Must be greater than 0.');
      }
      params.category_id = categoryId;
    }

    const response = await apiClient.get('/api/dhikr/random', { params });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get specific dhikr/dua by ID
 * @param {number} dhikrId - Unique identifier of the dhikr
 * @returns {Promise<Object>} Dhikr details
 */
export async function getDhikrById(dhikrId) {
  try {
    if (!dhikrId || dhikrId < 1) {
      throw new Error('Invalid dhikr ID. Must be greater than 0.');
    }

    const response = await apiClient.get(`/api/dhikr/${dhikrId}`);
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get all available dhikr/dua categories
 * @returns {Promise<Object>} List of categories with counts
 */
export async function getDhikrCategories() {
  try {
    const response = await apiClient.get('/api/dhikr/categories');
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Search dhikr/dua by text content
 * @param {Object} params - Search parameters
 * @param {string} params.q - Search term (min 2 characters)
 * @param {string} [params.language='both'] - Search language: 'arabic', 'english', or 'both'
 * @param {number} [params.limit=20] - Maximum results (max 100)
 * @param {number} [params.offset=0] - Pagination offset
 * @returns {Promise<Object>} List of matching dhikr
 */
export async function searchDhikr(params) {
  try {
    const { q, language = 'both', limit = 20, offset = 0 } = params;

    if (!q || q.length < 2) {
      throw new Error('Search query must be at least 2 characters long.');
    }
    if (!['arabic', 'english', 'both'].includes(language)) {
      throw new Error('Language must be "arabic", "english", or "both".');
    }
    if (limit > 100) {
      throw new Error('Limit cannot exceed 100.');
    }
    if (offset < 0) {
      throw new Error('Offset must be non-negative.');
    }

    const response = await apiClient.get('/api/dhikr/search', {
      params: { q, language, limit, offset }
    });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get morning adhkar (azkar al-sabah)
 * @param {number} [limit=10] - Number of adhkar to return
 * @returns {Promise<Object>} List of morning dhikr
 */
export async function getMorningDhikr(limit = 10) {
  try {
    if (limit < 1 || limit > 100) {
      throw new Error('Limit must be between 1 and 100.');
    }

    const response = await apiClient.get('/api/dhikr/morning', {
      params: { limit }
    });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Get evening adhkar (azkar al-masa)
 * @param {number} [limit=10] - Number of adhkar to return
 * @returns {Promise<Object>} List of evening dhikr
 */
export async function getEveningDhikr(limit = 10) {
  try {
    if (limit < 1 || limit > 100) {
      throw new Error('Limit must be between 1 and 100.');
    }

    const response = await apiClient.get('/api/dhikr/evening', {
      params: { limit }
    });
    return response;
  } catch (error) {
    throw error;
  }
}

/**
 * Search dhikr in Arabic only
 * @param {string} searchText - Arabic text to search
 * @param {number} [limit=20] - Maximum results
 * @returns {Promise<Object>} Search results
 */
export async function searchArabicDhikr(searchText, limit = 20) {
  return searchDhikr({
    q: searchText,
    language: 'arabic',
    limit
  });
}

/**
 * Search dhikr in English only
 * @param {string} searchText - English text to search
 * @param {number} [limit=20] - Maximum results
 * @returns {Promise<Object>} Search results
 */
export async function searchEnglishDhikr(searchText, limit = 20) {
  return searchDhikr({
    q: searchText,
    language: 'english',
    limit
  });
}

/**
 * Get dhikr for specific time of day
 * @param {string} timeOfDay - 'morning', 'evening', or 'general'
 * @param {number} [limit=10] - Number of dhikr to return
 * @returns {Promise<Object>} Dhikr for the specified time
 */
export async function getDhikrByTimeOfDay(timeOfDay, limit = 10) {
  try {
    const timeMap = {
      morning: getMorningDhikr,
      evening: getEveningDhikr,
      general: getRandomDhikr
    };

    const fetchFunction = timeMap[timeOfDay.toLowerCase()];
    if (!fetchFunction) {
      throw new Error('Invalid time of day. Must be "morning", "evening", or "general".');
    }

    return await fetchFunction(limit);
  } catch (error) {
    throw error;
  }
}

/**
 * Get dhikr appropriate for current time
 * @param {number} [limit=10] - Number of dhikr to return
 * @returns {Promise<Object>} Dhikr for current time
 */
export async function getDhikrForCurrentTime(limit = 10) {
  try {
    const currentHour = new Date().getHours();

    // Morning: 4 AM - 12 PM
    if (currentHour >= 4 && currentHour < 12) {
      return await getMorningDhikr(limit);
    }
    // Evening: 3 PM - 8 PM
    else if (currentHour >= 15 && currentHour < 20) {
      return await getEveningDhikr(limit);
    }
    // Other times: random/general dhikr
    else {
      return await getRandomDhikr();
    }
  } catch (error) {
    throw error;
  }
}

/**
 * Get all dhikr from multiple categories
 * @param {Array<number>} categoryIds - Array of category IDs
 * @param {number} [limit=20] - Results per category
 * @returns {Promise<Array>} Array of dhikr from all categories
 */
export async function getDhikrFromMultipleCategories(categoryIds, limit = 20) {
  try {
    if (!Array.isArray(categoryIds) || categoryIds.length === 0) {
      throw new Error('Category IDs must be a non-empty array.');
    }

    const promises = categoryIds.map(categoryId =>
      getDhikrByCategory(categoryId, limit).catch(err => ({
        error: true,
        categoryId,
        message: err.message
      }))
    );

    const results = await Promise.all(promises);
    return results;
  } catch (error) {
    throw error;
  }
}

/**
 * Transform dhikr data for frontend display
 * @param {Object} dhikr - Dhikr data from backend
 * @returns {Object} Transformed dhikr data
 */
export function transformDhikr(dhikr) {
  return {
    id: dhikr.id || dhikr.dhikr_id,
    arabic: dhikr.text_arabic || dhikr.dhikr_ar,
    english: dhikr.text_english || dhikr.dhikr_en,
    transliteration: dhikr.transliteration,
    translation: dhikr.translation || dhikr.english,
    reference: dhikr.reference || dhikr.source,
    category: dhikr.category || dhikr.category_name,
    categoryId: dhikr.category_id,
    repetitions: dhikr.repetitions || dhikr.count || 1,
    benefits: dhikr.benefits || dhikr.virtue,
    timeOfDay: dhikr.time_of_day || dhikr.timing
  };
}

/**
 * Get dhikr statistics for a category
 * @param {number} categoryId - Category ID
 * @returns {Promise<Object>} Category statistics
 */
export async function getCategoryStatistics(categoryId) {
  try {
    const dhikrList = await getDhikrByCategory(categoryId, 100, 0);

    if (!dhikrList || !dhikrList.results) {
      return null;
    }

    return {
      categoryId,
      totalCount: dhikrList.total || dhikrList.results.length,
      averageRepetitions:
        dhikrList.results.reduce((sum, d) => sum + (d.repetitions || 1), 0) /
        dhikrList.results.length
    };
  } catch (error) {
    throw error;
  }
}

/**
 * Create a dhikr routine for a specific time period
 * @param {string} period - 'daily', 'morning', 'evening', 'weekly'
 * @returns {Promise<Object>} Dhikr routine
 */
export async function createDhikrRoutine(period) {
  try {
    const routines = {
      daily: async () => {
        const morning = await getMorningDhikr(5);
        const evening = await getEveningDhikr(5);
        return { morning, evening };
      },
      morning: () => getMorningDhikr(10),
      evening: () => getEveningDhikr(10),
      weekly: async () => {
        const categories = await getDhikrCategories();
        if (categories && categories.categories) {
          const randomCategory =
            categories.categories[
              Math.floor(Math.random() * categories.categories.length)
            ];
          return await getDhikrByCategory(randomCategory.id, 7);
        }
        return null;
      }
    };

    const routineFunction = routines[period.toLowerCase()];
    if (!routineFunction) {
      throw new Error(
        'Invalid period. Must be "daily", "morning", "evening", or "weekly".'
      );
    }

    return await routineFunction();
  } catch (error) {
    throw error;
  }
}

// Export all functions as default object
export default {
  getDailyDhikr,
  getDhikrByCategory,
  getRandomDhikr,
  getDhikrById,
  getDhikrCategories,
  searchDhikr,
  getMorningDhikr,
  getEveningDhikr,
  searchArabicDhikr,
  searchEnglishDhikr,
  getDhikrByTimeOfDay,
  getDhikrForCurrentTime,
  getDhikrFromMultipleCategories,
  transformDhikr,
  getCategoryStatistics,
  createDhikrRoutine
};

File: backend/app/services/dhikr_service.py
"""
Dhikr Service - Business logic for dhikr and dua operations
"""

from typing import List, Dict, Any, Optional
from datetime import datetime, date
from ..database import execute_query, execute_query_single


class DhikrService:
    """Service class for dhikr and dua related business logic"""

    @staticmethod
    async def get_daily_dhikr(category_id: int = 1) -> Optional[Dict[str, Any]]:
        """
        Get a random dhikr for daily practice

        Args:
            category_id: Category of dhikr (default: general)

        Returns:
            Random dhikr from category
        """
        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "category_id" = $1
            ORDER BY (dhikr_id % 365) = EXTRACT(DOY FROM NOW())::int DESC,
                     dhikr_id
            LIMIT 1
        """

        return await execute_query_single(sql, category_id)

    @staticmethod
    async def get_dhikr_by_time_of_day() -> List[Dict[str, Any]]:
        """
        Get appropriate dhikr based on current time of day

        Returns:
            List of time-appropriate dhikr
        """
        now = datetime.now()
        hour = now.hour

        # Determine time category
        # Morning: 5am - 12pm (category 1)
        # Evening: 5pm - 8pm (category 2)
        # Night: 8pm - 5am (category 3)
        # General: 12pm - 5pm (category 4)

        if 5 <= hour < 12:
            category_id = 1  # Morning adhkar
        elif 17 <= hour < 20:
            category_id = 2  # Evening adhkar
        elif hour >= 20 or hour < 5:
            category_id = 3  # Night adhkar
        else:
            category_id = 4  # General adhkar

        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "category_id" = $1
            ORDER BY "dhikr_id" ASC
            LIMIT 10
        """

        return await execute_query(sql, category_id)

    @staticmethod
    async def get_dhikr_by_category(
        category_id: int, limit: int = 20, offset: int = 0
    ) -> Dict[str, Any]:
        """
        Get all dhikr from a specific category with pagination

        Args:
            category_id: Category ID
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Dhikr list with metadata
        """
        # Get total count
        count_sql = """
            SELECT COUNT(*) as total
            FROM "dhikr"
            WHERE "category_id" = $1
        """
        count_result = await execute_query_single(count_sql, category_id)
        total = count_result.get("total", 0) if count_result else 0

        # Get dhikr
        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "category_id" = $1
            ORDER BY "dhikr_id" ASC
            LIMIT $2 OFFSET $3
        """

        dhikr_list = await execute_query(sql, category_id, limit, offset)

        return {
            "results": dhikr_list,
            "total": total,
            "limit": limit,
            "offset": offset,
            "has_more": (offset + len(dhikr_list)) < total,
        }

    @staticmethod
    async def search_dhikr(
        query: str, language: str = "both", limit: int = 20, offset: int = 0
    ) -> Dict[str, Any]:
        """
        Search dhikr by text content

        Args:
            query: Search term
            language: 'arabic', 'english', or 'both'
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Search results with metadata
        """
        search_pattern = f"%{query}%"

        # Build query based on language
        if language == "arabic":
            sql = """
                SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                       "benefits_ar", "benefits_en", "reference"
                FROM "dhikr"
                WHERE "text_ar" ILIKE $1 OR "benefits_ar" ILIKE $1
                ORDER BY "dhikr_id" ASC
            """
            params = [search_pattern]
        elif language == "english":
            sql = """
                SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                       "benefits_ar", "benefits_en", "reference"
                FROM "dhikr"
                WHERE "text_en" ILIKE $1 OR "benefits_en" ILIKE $1
                ORDER BY "dhikr_id" ASC
            """
            params = [search_pattern]
        else:  # both
            sql = """
                SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                       "benefits_ar", "benefits_en", "reference"
                FROM "dhikr"
                WHERE "text_ar" ILIKE $1 OR "text_en" ILIKE $1
                   OR "benefits_ar" ILIKE $1 OR "benefits_en" ILIKE $1
                ORDER BY "dhikr_id" ASC
            """
            params = [search_pattern]

        # Get count
        count_sql = sql.replace(
            'SELECT "dhikr_id", "category_id", "text_ar", "text_en", "benefits_ar", "benefits_en", "reference"',
            "SELECT COUNT(*) as total",
        ).replace('ORDER BY "dhikr_id" ASC', "")

        count_result = await execute_query_single(count_sql, *params)
        total = count_result.get("total", 0) if count_result else 0

        # Add pagination
        sql += f" LIMIT ${len(params) + 1} OFFSET ${len(params) + 2}"
        params.extend([limit, offset])

        results = await execute_query(sql, *params)

        return {
            "results": results,
            "total": total,
            "limit": limit,
            "offset": offset,
            "has_more": (offset + len(results)) < total,
            "query": query,
            "language": language,
        }

    @staticmethod
    async def get_dhikr_with_benefits() -> List[Dict[str, Any]]:
        """
        Get all dhikr that have documented benefits

        Returns:
            List of dhikr with benefits
        """
        sql = """
            SELECT d."dhikr_id", d."category_id", d."text_ar", d."text_en",
                   d."benefits_ar", d."benefits_en", d."reference",
                   c."name_ar" as category_name_ar, c."name_en" as category_name_en
            FROM "dhikr" d
            JOIN "categories" c ON d."category_id" = c."category_id"
            WHERE (d."benefits_ar" IS NOT NULL AND d."benefits_ar" != '')
               OR (d."benefits_en" IS NOT NULL AND d."benefits_en" != '')
            ORDER BY d."category_id" ASC, d."dhikr_id" ASC
        """

        return await execute_query(sql)

    @staticmethod
    async def get_dhikr_by_reference(reference: str) -> List[Dict[str, Any]]:
        """
        Get dhikr from a specific hadith or Quran reference

        Args:
            reference: Reference source (e.g., "Sahih Muslim", "Quran 2:255")

        Returns:
            List of dhikr with matching reference
        """
        search_pattern = f"%{reference}%"

        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "reference" ILIKE $1
            ORDER BY "dhikr_id" ASC
        """

        return await execute_query(sql, search_pattern)

    @staticmethod
    async def get_all_categories() -> List[Dict[str, Any]]:
        """
        Get all dhikr categories with counts

        Returns:
            List of categories with dhikr counts
        """
        sql = """
            SELECT c."category_id", c."name_ar", c."name_en",
                   COUNT(d."dhikr_id") as dhikr_count
            FROM "categories" c
            LEFT JOIN "dhikr" d ON c."category_id" = d."category_id"
            GROUP BY c."category_id", c."name_ar", c."name_en"
            ORDER BY c."category_id" ASC
        """

        return await execute_query(sql)

    @staticmethod
    async def get_morning_adhkar(limit: int = 10) -> List[Dict[str, Any]]:
        """
        Get morning adhkar (azkar al-sabah)

        Args:
            limit: Number of adhkar to return

        Returns:
            List of morning adhkar
        """
        # Assuming category_id 1 is for morning adhkar
        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "category_id" = 1
            ORDER BY "dhikr_id" ASC
            LIMIT $1
        """

        return await execute_query(sql, limit)

    @staticmethod
    async def get_evening_adhkar(limit: int = 10) -> List[Dict[str, Any]]:
        """
        Get evening adhkar (azkar al-masa)

        Args:
            limit: Number of adhkar to return

        Returns:
            List of evening adhkar
        """
        # Assuming category_id 2 is for evening adhkar
        sql = """
            SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                   "benefits_ar", "benefits_en", "reference"
            FROM "dhikr"
            WHERE "category_id" = 2
            ORDER BY "dhikr_id" ASC
            LIMIT $1
        """

        return await execute_query(sql, limit)

    @staticmethod
    async def get_random_dhikr(
        category_id: Optional[int] = None,
    ) -> Optional[Dict[str, Any]]:
        """
        Get a random dhikr, optionally from a specific category

        Args:
            category_id: Optional category filter

        Returns:
            Random dhikr
        """
        if category_id:
            sql = """
                SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                       "benefits_ar", "benefits_en", "reference"
                FROM "dhikr"
                WHERE "category_id" = $1
                ORDER BY (dhikr_id % 365) = EXTRACT(DOY FROM NOW())::int DESC,
                         dhikr_id
                LIMIT 1
            """
            return await execute_query_single(sql, category_id)
        else:
            sql = """
                SELECT "dhikr_id", "category_id", "text_ar", "text_en",
                       "benefits_ar", "benefits_en", "reference"
                FROM "dhikr"
                ORDER BY (dhikr_id % 365) = EXTRACT(DOY FROM NOW())::int DESC,
                         dhikr_id
                LIMIT 1
            """
            return await execute_query_single(sql)

    @staticmethod
    async def get_dhikr_collection(dhikr_ids: List[int]) -> List[Dict[str, Any]]:
        """
        Get multiple dhikr by their IDs (useful for saved collections)

        Args:
            dhikr_ids: List of dhikr IDs to retrieve

        Returns:
            List of dhikr matching the IDs
        """
        if not dhikr_ids:
            return []

        # Create parameter placeholders
        placeholders = ",".join([f"${i+1}" for i in range(len(dhikr_ids))])

        sql = f"""
            SELECT d."dhikr_id", d."category_id", d."text_ar", d."text_en",
                   d."benefits_ar", d."benefits_en", d."reference",
                   c."name_ar" as category_name_ar, c."name_en" as category_name_en
            FROM "dhikr" d
            JOIN "categories" c ON d."category_id" = c."category_id"
            WHERE d."dhikr_id" IN ({placeholders})
            ORDER BY d."dhikr_id" ASC
        """

        return await execute_query(sql, *dhikr_ids)

    @staticmethod
    async def get_dhikr_statistics() -> Dict[str, Any]:
        """
        Get overall dhikr statistics

        Returns:
            Statistics about dhikr in database
        """
        sql = """
            SELECT
                COUNT(*) as total_dhikr,
                COUNT(DISTINCT "category_id") as total_categories,
                COUNT(CASE WHEN "benefits_ar" IS NOT NULL THEN 1 END) as dhikr_with_benefits,
                COUNT(CASE WHEN "reference" IS NOT NULL THEN 1 END) as dhikr_with_references
            FROM "dhikr"
        """

        stats = await execute_query_single(sql)

        # Get category breakdown
        category_sql = """
            SELECT c."name_en", c."name_ar", COUNT(d."dhikr_id") as count
            FROM "categories" c
            LEFT JOIN "dhikr" d ON c."category_id" = d."category_id"
            GROUP BY c."category_id", c."name_en", c."name_ar"
            ORDER BY count DESC
        """
        category_breakdown = await execute_query(category_sql)

        return {"overview": stats, "category_breakdown": category_breakdown}

    @staticmethod
    def get_dhikr_of_the_day_deterministic(date_obj: Optional[date] = None) -> int:
        """
        Get a deterministic dhikr ID based on the date
        (same dhikr for the same date across all users)

        Args:
            date_obj: Date for calculation (defaults to today)

        Returns:
            Dhikr ID that should be featured for this date
        """
        if date_obj is None:
            date_obj = date.today()

        # Use day of year to select dhikr (cycles through year)
        day_of_year = date_obj.timetuple().tm_yday

        # Assuming we have at least 365 dhikr entries
        # This creates a deterministic selection based on date
        # Can be adjusted based on actual dhikr count
        return (day_of_year % 100) + 1

    @staticmethod
    async def get_featured_dhikr() -> Optional[Dict[str, Any]]:
        """
        Get the featured dhikr of the day (deterministic based on date)

        Returns:
            Featured dhikr for today
        """
        dhikr_id = DhikrService.get_dhikr_of_the_day_deterministic()

        sql = """
            SELECT d."dhikr_id", d."category_id", d."text_ar", d."text_en",
                   d."benefits_ar", d."benefits_en", d."reference",
                   c."name_ar" as category_name_ar, c."name_en" as category_name_en
            FROM "dhikr" d
            JOIN "categories" c ON d."category_id" = c."category_id"
            WHERE d."dhikr_id" = $1
        """

        result = await execute_query_single(sql, dhikr_id)

        # If specific dhikr not found, return a random one
        if not result:
            return await DhikrService.get_random_dhikr()

        return result
